src/samples/http/server-sse.c

```diff
@@ -0,0 +1,295 @@
+ * type something
+ *
+ * To see the usage help, -h or --help.
+ */
+
+
+#include <getopt.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "soletta.h"
+#include "sol-http.h"
+#include "sol-http-server.h"
+#include "sol-util.h"
+#include "sol-util-file.h"
+
+#define HTML_FILE \
+    "<!DOCTYPE html>" \
+    "<html>" \
+    "<body>" \
+    "<h1>Getting server updates</h1>" \
+    "<div id=\"result\"></div>" \
+    "<script>" \
+    "if(typeof(EventSource) !== \"undefined\") {" \
+    "var source = new EventSource(\"http://%.*s:%d/events\");" \
+    "source.onmessage = function(event) {" \
+    "document.getElementById(\"result\").innerHTML += event.data + \"<br>\";" \
+    "};" \
+    "} else {" \
+    "document.getElementById(\"result\").innerHTML = \"Sorry, your browser does not support server-sent events...\";" \
+    "}" \
+    "</script>" \
+    "</body>" \
+    "</html>"
+
+
+static struct sol_http_server *server;
+static struct sol_fd *stdin_watch;
+static struct sol_buffer value;
+static struct sol_ptr_vector responses = SOL_PTR_VECTOR_INIT;
+static int port = 8080;
+
+static bool
+on_stdin(void *data, int fd, uint32_t flags)
+{
+    uint16_t i;
+    struct sol_http_response_progressive *sse;
+    static const struct sol_str_slice begin = {
+        .len = sizeof("data: ") - 1,
+        .data = (const char *)"data: "
+    };
+    static const struct sol_str_slice end = {
+        .len = sizeof("\n\n") - 1,
+        .data = (const char *)"\n\n"
+    };
+
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP)) {
+        fprintf(stderr, "ERROR: Something wrong happened with file descriptor: %d\n", fd);
+        goto err;
+    }
+
+    if (flags & SOL_FD_FLAGS_IN) {
+        int err;
+
+        value.used = 0;
+        /* this will loop trying to read as much data as possible to buffer. */
+        err = sol_util_load_file_fd_buffer(fd, &value);
+        if (err < 0) {
+            fprintf(stderr, "ERROR: failed to read from stdin: %s\n",
+                sol_util_strerrora(-err));
+            goto err;
+        }
+
+        if (value.used == 0) {
+            /* no data usually means ^D on the terminal, quit the application */
+            printf("no data on stdin, quitting.\n");
+            sol_quit();
+        } else {
+            err = sol_buffer_insert_slice(&value, 0, begin);
+            if (err < 0) {
+                fprintf(stderr, "ERROR: failed to append the data prefix: %s\n",
+                    sol_util_strerrora(-err));
+                goto err;
+            }
+
+            err = sol_buffer_append_slice(&value, end);
+            if (err < 0) {
+                fprintf(stderr, "ERROR: Could not append data to buffer: %s\n",
+                    sol_util_strerrora(-err));
+                goto err;
+            }
+        }
+    }
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&responses, sse, i)
+        sol_http_response_progressive_feed(sse, sol_buffer_get_slice(&value));
+
```

Comment:
my bad, there is a `sol_buffer_fini()` in shutdown function. But as I see there is no reason why this buffer should be global.